### PR TITLE
Add handling for testing against deployments

### PR DESCRIPTION
**Pull request by ijjk**
This adds a new test mode `next-deploy` which allows testing against deployments using the Vercel CLI. After running these tests they uncovered some bugs we need to correct before fully enabling. Patching the uncovered issues will be done in follow-up PRs and then after resolved this will be enabled to run after new publishes. 

Tests that uncovered bugs to patch in follow-ups:

- [x] test/e2e/getserversideprops/test/index.test.ts (req.url normalizing)
- [ ] test/e2e/i18n-api-support/index.test.ts (locale prefixed API routes matching)
- [x] test/e2e/prerender.test.ts (/_next/data/build-id/ does not 404)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 2485de244d49af5462c0f4581f0574602b703e4c

`yarn testheadless test/e2e/prerender.test.ts` 
- Prerender > outputs dataRoutes in routes-manifest correctly

<details>
<summary>Expand output</summary>

  ● Prerender › outputs dataRoutes in routes-manifest correctly

    expect(received).toEqual(expected) // deep equality

    - Expected  - 4
    + Received  + 0

    @@ -111,14 +111,10 @@
          "routeKeys": Object {
            "slug": "slug",
          },
        },
        Object {
    -     "dataRouteRegex": "^\\/_next\\/data\\/LroODQLvZJ85faIlfkWhX\\/index\\/index.json$",
    -     "page": "/index",
    -   },
    -   Object {
          "dataRouteRegex": "^\\/_next\\/data\\/LroODQLvZJ85faIlfkWhX\\/lang\\/([^\\/]+?)\\/about\\.json$",
          "namedDataRouteRegex": "^/_next/data/LroODQLvZJ85faIlfkWhX/lang/(?<lang>[^/]+?)/about\\.json$",
          "page": "/lang/[lang]/about",
          "routeKeys": Object {
            "lang": "lang",

      1204 |           }
      1205 |
    > 1206 |           expect(dataRoutes).toEqual([
           |                              ^
      1207 |             {
      1208 |               dataRouteRegex: normalizeRegEx(
      1209 |                 `^\\/_next\\/data\\/${escapeRegex(next.buildId)}\\/index.json$`

      at Object.<anonymous> (e2e/prerender.test.ts:1206:30)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 2c73a178cc4ba2add475b1685842a94f8353dc1f

`yarn testheadless test/e2e/basepath.test.ts` 
- basePath > should prefetch pages correctly in viewport with <Link>
- basePath > should navigate to index page with getStaticProps

<details>
<summary>Expand output</summary>

  ● basePath › should prefetch pages correctly in viewport with <Link>

    TIMED OUT: yes

    undefined

      498 |
      499 |   if (hardError) {
    > 500 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content)
          |           ^
      501 |   }
      502 |   return false
      503 | }

      at Object.check (lib/next-test-utils.js:500:11)
      at Object.<anonymous> (e2e/basepath.test.ts:208:9)

  ● basePath › should navigate to index page with getStaticProps

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 0

      Array [
        "/docs/gsp.json",
        "/docs/index.json",
    -   "/docs/index/index.json",
      ]

      436 |             new URL(href).pathname.replace(/\/_next\/data\/[^/]+/, '')
      437 |           )
    > 438 |         ).toEqual([
          |           ^
      439 |           `${basePath}/gsp.json`,
      440 |           `${basePath}/index.json`,
      441 |           `${basePath}/index/index.json`,

      at Object.<anonymous> (e2e/basepath.test.ts:438:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| buildDuration | 16.1s | 15.9s | -196ms |
| buildDurationCached | 6.1s | 6.2s | ⚠️ +113ms |
| nodeModulesSize | 485 MB | 485 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.23 | 3.207 | -0.02 |
| / avg req/sec | 773.89 | 779.5 | +5.61 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.404 | 1.37 | -0.03 |
| /error-in-render avg req/sec | 1780.4 | 1824.66 | +44.26 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.5 kB | 28.5 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.74 kB | 5.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 461 B | ✓ |
| Overall change | 461 B | 461 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| buildDuration | 18.6s | 18.3s | -278ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +7ms |
| nodeModulesSize | 485 MB | 485 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.248 | 3.195 | -0.05 |
| / avg req/sec | 769.62 | 782.42 | +12.8 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.36 | 1.343 | -0.02 |
| /error-in-render avg req/sec | 1838.53 | 1861.87 | +23.34 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.9 kB | 28.9 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.8 kB | 72.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.78 kB | 5.78 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| _buildManifest.js gzip | 457 B | 457 B | ✓ |
| Overall change | 457 B | 457 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js tests/e2e-deploy | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 2c73a178cc4ba2add475b1685842a94f8353dc1f

**Review by timneutkens — APPROVED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 6ff4029816050258c9de910717f6e589f7d27418

`yarn testheadless test/e2e/prerender.test.ts` 
- Prerender > should always show fallback for page not in getStaticPaths

<details>
<summary>Expand output</summary>

  ● Prerender › should always show fallback for page not in getStaticPaths

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      941 |         const html = await renderViaHTTP(next.url, '/blog/post-321')
      942 |         const $ = cheerio.load(html)
    > 943 |         expect(JSON.parse($('#__NEXT_DATA__').text()).isFallback).toBe(true)
          |                                                                   ^
      944 |
      945 |         // make another request to ensure it still is
      946 |         const html2 = await renderViaHTTP(next.url, '/blog/post-321')

      at Object.<anonymous> (e2e/prerender.test.ts:943:67)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/readme.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ You can set-up a new test using `yarn new-test` which will start from a template
 
 ## Test Types in Next.js
 
-- e2e: These tests will run against `next dev` and `next start`
+- e2e: These tests will run against `next dev`, `next start`, and deployed to Vercel
```
Comment (styfle) [inline test/readme.md:9]: Would it make sense to also run these tests agains “vc dev”?

Comment (ijjk) [reply test/readme.md:9]: Possibly although since `vc dev` proxies to `next dev` now I'm not sure the specifics we would be testing 🤔

**Pull request merged**